lib/ClangImporter/ClangModuleDependencyScanner.cpp

```diff
@@ -451,86 +451,98 @@ ClangImporter::getModuleDependencies(Identifier moduleName,
+        clangScanningTool, dependencies->ModuleGraph,
+        cache.getModuleOutputPath(),
+        [&cache](StringRef path) {
+          return cache.getScanService().remapPath(path);
+        });
+    cache.recordDependencies(bridgedDeps);
 
-  // Determine the command-line arguments for dependency scanning.
-  std::vector<std::string> commandLineArgs =
-      getClangDepScanningInvocationArguments(ctx, StringRef(bridgingHeader));
-  auto optionalWorkingDir = computeClangWorkingDirectory(commandLineArgs, ctx);
-  if (!optionalWorkingDir) {
-    ctx.Diags.diagnose(SourceLoc(), diag::clang_dependency_scan_error,
-                       "Missing '-working-directory' argument");
-    return true;
-  }
-  std::string workingDir = *optionalWorkingDir;
+    // Record dependencies for the source files the bridging header includes.
+    for (const auto &fileDep : dependencies->FileDeps)
+      targetModule.addHeaderSourceFile(fileDep);
 
-  auto moduleCachePath = getModuleCachePathFromClang(getClangInstance());
-  auto lookupModuleOutput =
-      [moduleCachePath](const ModuleID &MID,
-                        ModuleOutputKind MOK) -> std::string {
-    return moduleCacheRelativeLookupModuleOutput(MID, MOK, moduleCachePath);
+    // ... and all module dependencies.
+    llvm::StringSet<> alreadyAddedModules;
+    for (const auto &moduleDep : dependencies->ClangModuleDeps)
+      targetModule.addHeaderInputModuleDependency(moduleDep.ModuleName,
+                                                  alreadyAddedModules);
+
+    return dependencies;
   };
 
-  auto clangModuleDependencies =
-      clangScanningTool.getTranslationUnitDependencies(
-          commandLineArgs, workingDir, cache.getAlreadySeenClangModules(),
-          lookupModuleOutput);
-  if (!clangModuleDependencies) {
-    // FIXME: Route this to a normal diagnostic.
-    llvm::logAllUnhandledErrors(clangModuleDependencies.takeError(), llvm::errs());
-    return true;
+  // - Textual module dependencies require us to process their bridging header.
+  // - Binary module dependnecies may have arbitrary header inputs.
+  if (targetModule.isTextualSwiftModule() &&
+      !targetModule.getBridgingHeader()->empty()) {
+    auto clangModuleDependencies =
+        scanHeaderDependencies(*targetModule.getBridgingHeader());
+    if (!clangModuleDependencies)
+      return true;
+    if (auto TreeID = clangModuleDependencies->IncludeTreeID)
+      targetModule.addBridgingHeaderIncludeTree(*TreeID);
+    recordBridgingHeaderOptions(targetModule, *clangModuleDependencies);
+    // Update the cache with the new information for the module.
+    cache.updateDependency(moduleID, targetModule);
+  } else if (targetModule.isSwiftBinaryModule()) {
+    auto swiftBinaryDeps = targetModule.getAsSwiftBinaryModule();
+    if (!swiftBinaryDeps->headerImports.empty()) {
+      for (const auto &headerPath : swiftBinaryDeps->headerImports) {
+        auto clangModuleDependencies = scanHeaderDependencies(headerPath);
```

Comment:
I don't feel like this is correct. You can only `scanHeaderDependencies` on the one bridging header as other headers might not even be standalone and missing other includes. That scan on the bridging header will return all the dependencies that covers all other headers.

That includes CAS case, where the include tree can only be created for entire PCH compilation, not each individual file.

Reply:
I believe that in practice, this is what we will get. i.e. in practice a given binary module we load will only ever have one entry in the `headerImports` which corresponds to its bridging header. It's just that the way it is represented in a serialized binary Swift module is that there can be an arbitrary number of such headers, so I wrote the code to handle it as such.

Reply:
I see now. Also why you need two fields for source/header. You want to keep track of the bridging header so you can run a dependency scanning here not including other headers. LGTM but maybe keep haederImport to be one element.

Reply:
I've removed the additional headerSources tracking. Thanks!

Reply:
hmm, I don' think that is what I meant. I think you might actually need to track them separately but this loop is definitely not needed since you should really only scan one header here.

Reply:
Ah got it. Brought back source file tracking and replaced this loop with just a scan of a single header.